src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,476 @@
+            {
+                WriteMetadataReference(writer, reference);
+            }
+            writer.WriteArrayEnd();
+            writer.WriteObjectEnd();
+
+            void writeToolsVersions()
+            {
+                writer.WriteKey("toolsVersions");
+                writer.WriteObjectStart();
+                if (0 != (options & DeterministicKeyOptions.IgnoreToolVersions))
+                {
+                    writer.WriteObjectEnd();
+                    return;
+                }
+
+                var compilerVersion = typeof(Compilation).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
+                writer.Write("compilerVersion", compilerVersion);
+
+                var runtimeVersion = typeof(object).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
+                writer.Write("runtimeVersion", runtimeVersion);
+
+                writer.Write("framework", RuntimeInformation.FrameworkDescription);
+                writer.Write("os", RuntimeInformation.OSDescription);
+
+                writer.WriteObjectEnd();
+            }
+        }
+
+        private void WriteSyntaxTree(JsonWriter writer, SyntaxTree syntaxTree, DeterministicKeyOptions options)
+        {
+            writer.WriteObjectStart();
+            WriteFileName(writer, "fileName", syntaxTree.FilePath, options);
+            writer.WriteKey("text");
+            WriteSourceText(writer, syntaxTree.GetText());
+            writer.WriteKey("parseOptions");
+            WriteParseOptions(writer, syntaxTree.Options);
+            writer.WriteObjectEnd();
+        }
+
+        private void WriteSourceText(JsonWriter writer, SourceText? sourceText)
+        {
+            if (sourceText is null)
+            {
+                return;
+            }
+
+            writer.WriteObjectStart();
+            WriteByteArrayValue(writer, "checksum", sourceText.GetChecksum());
+            writer.Write("checksumAlgorithm", sourceText.ChecksumAlgorithm);
+            writer.Write("encoding", sourceText.Encoding?.EncodingName);
+            writer.WriteObjectEnd();
+        }
+
+        internal void WriteMetadataReference(JsonWriter writer, MetadataReference reference)
+        {
+            writer.WriteObjectStart();
+            if (reference is PortableExecutableReference peReference)
+            {
+                ModuleMetadata moduleMetadata;
+                switch (peReference.GetMetadata())
+                {
+                    case AssemblyMetadata assemblyMetadata:
+                        {
+                            if (assemblyMetadata.GetModules() is { Length: 1 } modules)
+                            {
+                                moduleMetadata = modules[0];
+                            }
+                            else
+                            {
+                                throw new InvalidOperationException();
```

Comment:
Is this a TODO or do we just not support netmodules here? Should we have a comment to that effect?

Reply:
Yes I need to fix this up before merge.